### ISSUE TITLE
Shrev setup example doesn't compile

### ISSUE THREAD

**Issue by Nashenas88** [slide-rs/specs#473]
Using this code:
```rust
extern crate specs;
extern crate shrev;

use specs::{System, Read, Resources};
use shrev::{EventChannel, ReaderId};

struct Event {
    s: String,
}

struct MySystem {
    reader: Option<ReaderId<Event>>,
}

impl<'a> System<'a> for MySystem {
    type SystemData = (Read<'a, EventChannel<Event>>);
    fn run(&mut self, events: Self::SystemData) {
        for event in events.read(&mut self.reader.as_mut().unwrap()) {
            println!("{:?}", event.s);
        }
    }

    fn setup(&mut self, res: &mut Resources) {
        use specs::SystemData;
        Self::SystemData::setup(res);
        self.reader = Some(res.fetch::<EventChannel<Event>>().register_reader());
    }
}
```
And this as my dependencies in `Cargo.toml`:
```
[dependencies]
specs = "0.12"
shrev = "1.0"
```

I get the following error from `cargo check`:
```
error[E0596]: cannot borrow immutable borrowed content as mutable
  --> src/lib.rs:26:28
   |
26 |         self.reader = Some(res.fetch::<EventChannel<Event>>().register_reader());
   |                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot borrow as mutable

error: aborting due to previous error

For more information about this error, try `rustc --explain E0596`.
```

I attempted to rewrite the two lines of `setup` as 
```
        let mut x: specs::shred::Fetch<'_, EventChannel<Event>> = res.fetch::<EventChannel<Event>>();
        self.reader = Some(x.register_reader());
```
But then I get this error:
```
error[E0596]: cannot borrow immutable borrowed content as mutable
  --> src/lib.rs:27:28
   |
27 |         self.reader = Some(x.register_reader());
   |                            ^ cannot borrow as mutable

warning: variable does not need to be mutable
  --> src/lib.rs:26:13
   |
26 |         let mut x: specs::shred::Fetch<'_, EventChannel<Event>> = res.fetch::<EventChannel<Event>>();
   |             ----^
   |             |
   |             help: remove this `mut`
   |
   = note: #[warn(unused_mut)] on by default

error: aborting due to previous error

For more information about this error, try `rustc --explain E0596`.
```

So the book has an invalid example at this point. (Section 7: Setup).

**Comment by Nashenas88** [comment]
The issue seems to be fixed if I replace the last two lines of `setup` with:
```rust
        let mut channel: Write<'_, EventChannel<Event>> = SystemData::fetch(res);
        self.reader = Some(channel.register_reader());
```

Does this make sense given that `res` (`Self::SystemData`)'s type is `Read<'a, EventChannel<Event>>`? Why is it valid to use `Write` here?

**Comment by torkleyy** [comment]
Yes, the `register_reader` call needs a mutable event channel.

> Why is it valid [..]?

You can fetch whatever you want in `setup`, it's completely independent from your `SystemData` type.

You can also just use `fetch_mut` instead of `fetch` in your original example.

**Comment by torkleyy** [comment]
The fixed code sample in #474 should work, please tell me if not.
